### ISSUE TITLE
It is possible to create excessively huge arrays in VRL

### ISSUE THREAD

**Issue by FungusHumungus** [timberio/vector#6606]
In VRL running:

```
.thing[9223372036854775807] = 1
```

will cause VRL to attempt to allocate a huge array. This basically locks it up and will likely crash it eventually (I didn't wait long enough to find out for sure).

We should probably put a reasonable limit in here to prevent this from occurring.

This will become much more important when we allow variables to be used as array indexers.

**Comment by JeanMertz** [comment]
Thanks for creating the issue.

I don't consider this a "bug" by or in itself, as it was a conscious choice to pad the array with `null` values, but I do agree that it's something we want to take another look at if/when we start allowing dynamic indices.

I also think it's fair to add an upper limit to arrays, and have that be documented, as part of the "guard against slow programs" mantra.

**Comment by jszwedko** [comment]
An alternative could be some sort of sparse array implementation that wouldn't literally pad.

**Comment by JeanMertz** [comment]
> An alternative could be some sort of sparse array implementation that wouldn't literally pad.

Good idea. That would work for assignments to variables. However, for external event assignments that wouldn't help, since it has to serialize the sparse array back to a regular json-like array, at which point you still have the same issue.

**Comment by jszwedko** [comment]
> > An alternative could be some sort of sparse array implementation that wouldn't literally pad.
> 
> Good idea. That would work for assignments to variables. However, for external event assignments that wouldn't help, since it has to serialize the sparse array back to a regular json-like array, at which point you still have the same issue.

Aha, true. I would maybe just not allow assigning dynamic indexes outside of the current bounds of the array. Users could use `push, `append`, or `prepend` functions to add elements, along with methods for initializing arrays of given lengths.
